common/beerocks/bcl/include/bcl/network/file_descriptor_impl.h

```diff
@@ -0,0 +1,139 @@
+ * programming errors.
+ *
+ * This class will be aggregated by all classes modeling OS resources implementing the file
+ * descriptor interface.
+ */
+class FileDescriptorImpl : public FileDescriptor {
+public:
+    /**
+     * @brief Class constructor.
+     *
+     * @param fd File descriptor value
+     */
+    explicit FileDescriptorImpl(int fd) : m_fd(fd)
+    {
+        if (invalid_descriptor == fd) {
+            LOG(ERROR) << "Invalid file descriptor: " << strerror(errno);
+        }
+    }
+
+    /**
+     * @brief Copy constructor
+     *
+     * Delete copy constructor to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl(const FileDescriptorImpl &) = delete;
+
+    /**
+     * @brief Move constructor.
+     *
+     * A move constructor allows the resources owned by an rvalue object to be moved into an
+     * lvalue without creating its copy.
+     */
+    FileDescriptorImpl(FileDescriptorImpl &&obj)
+    {
+        this->m_fd = obj.m_fd;
+        obj.m_fd   = FileDescriptor::invalid_descriptor;
+    }
+
+    /**
+     * @brief Assignment operator
+     *
+     * Delete assignment operator to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl &operator=(const FileDescriptorImpl &) = delete;
+
+    /**
+     * @brief Move assignment operator
+     *
+     * The move assignment operator is used to transfer ownership of a file descriptor
+     */
+    FileDescriptorImpl &operator=(FileDescriptorImpl &&obj)
+    {
+        // Self-assignment detection
+        if (&obj == this) {
+            return *this;
+        }
+
+        // Release any resource we're holding
+        close();
+
+        // Transfer ownership of obj.m_fd to m_fd
+        this->m_fd = obj.m_fd;
+
+        return *this;
+    }
+
+    /**
+     * @brief Class destructor.
+     *
+     * Close socket on destructor (if it is still open)
+     */
+    ~FileDescriptorImpl() override { close(); }
+
+    /**
+     * @brief Gets file descriptor.
+     *
+     * File descriptor can be used for example in:
+     * - select(), poll() or epoll() to wait for events on this descriptor.
+     * - send() family of functions to send data through the socket connection.
+     * - recv() family of functions to receive data from the socket connection.
+     *
+     * @return File descriptor value.
+     */
+    int fd() override { return m_fd; }
+
+private:
+    /**
+     * File descriptor value
+     */
+    int m_fd;
+
+    /**
+     * @brief Closes file descriptor.
+     *
+     * If valid, closes file descriptor and then invalidates it.
+     *
+     * @return True on success and false otherwise (for example, if it was already closed)
+     */
+    bool close()
+    {
+        if (FileDescriptor::invalid_descriptor == m_fd) {
+            return false;
+        }
+
+        int rc = ::close(m_fd);
+        m_fd   = FileDescriptor::invalid_descriptor;
+
+        return (0 == rc);
```

Comment:
Inability to close something should not happen often. Error/warning message would be appropriate here.
Otherwise all users of `close()` need to check their return value.

Reply:
Fixed